hazelcast/src/main/java/com/hazelcast/util/collection/InflatableSet.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.util.collection;
+
+import java.io.Serializable;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Provides fast {@link Set} implementation for cases where items are known to not
+ * contain duplicates.
+ *
+ * It doesn't not call equals/hash methods on initial data insertion hence it avoids
+ * performance penalty in the case these methods are expensive. It also means it does
+ * not detect duplicates - it's a responsibility of a caller to make sure no duplicated
+ * entries are inserted.
+ *
+ * Once the initial load is done then caller should indicate it
+ * by calling {@link #close()}. It switches operational mode and if any item is
+ * inserted afterward then the Set will inflate - it copies its content into
+ * internal HashSet. This means duplicate detection is enabled, but obviously
+ * it ruins the initial performance gain. We are making a bet the Set will not be
+ * modified once it's closed. The Set will also inflate on {@link #contains(Object)}
+ * and {@link #containsAll(Collection)} calls to enable fast look-ups.
+ *
+ * Failing to call {@link #close()} means the InflatableSet is in state where
+ * a general contract of Set is broken - duplicates aren't detected. You should
+ * never pass unclosed InfltableSet to user code.
+ *
+ * It's intended to be use in cases where a contract mandates us to return Set,
+ * but we know our data contains not duplicates. It performs the best in cases
+ * biased towards sequential iterations.
+ *
+ *
+ * @param <T>
+ */
+public final class InflatableSet<T> extends AbstractSet<T> implements Set<T>, Serializable, Cloneable {
```

Comment:
Do you think we need to test serializability & de-serializability of `InflatableSet`?